gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/AccessControlServiceImpl.java

```diff
@@ -0,0 +1,197 @@
+    public boolean canAccessApiFromPortal(String apiId) {
+        ApiEntity apiEntity = apiService.findById(apiId);
+        return canAccessApiFromPortal(apiEntity);
+    }
+
+    private boolean canAccessPage(ApiEntity apiEntity, PageEntity pageEntity) {
+
+        if (!pageEntity.isPublished()) {
+            return false;
+        }
+
+        // in public mode
+        if (PUBLIC.equals(pageEntity.getVisibility())) {
+            return true;
+        }
+
+        // private mode
+        if (!isAuthenticated()) {
+            return false;
+        }
+
+        Set<AccessControlEntity> accessControls = pageEntity.getAccessControls();
+        if (accessControls == null || accessControls.isEmpty()) {
+            return true;
+        } else {
+
+            Set<GroupEntity> userGroups = groupService.findByUser(getAuthenticatedUsername());
+            Set<RoleEntity> contextualUserRoles = getContextualUserRoles(apiEntity);
+
+            return accessControls.stream()
+                .anyMatch(acl -> {
+                    if (AccessControlReferenceType.ROLE.name().equals(acl.getReferenceType())) {
+                        boolean roleMatched = contextualUserRoles.stream()
+                            .anyMatch((role) -> role.getId().equals(acl.getReferenceId()));
+                        return pageEntity.isExcludedAccessControls() ? !roleMatched : roleMatched;
+                    } else if (AccessControlReferenceType.GROUP.name().equals(acl.getReferenceType())) {
+                        boolean groupMatched = userGroups.stream().anyMatch((group) -> group.getId().equals(acl.getReferenceId()));
+                        return pageEntity.isExcludedAccessControls() ? !groupMatched : groupMatched;
+                    } else {
+                        logger.warn("ACL reference type [{}] not found", acl.getReferenceType());
+                    }
+                    return false;
+                });
+        }
+    }
+
+    @Override
+    public boolean canAccessPageFromPortal(PageEntity pageEntity) {
+        if (PageType.SYSTEM_FOLDER.name().equals(pageEntity.getType()) || PageType.MARKDOWN_TEMPLATE.name().equals(pageEntity.getType())) {
+            return false;
+        }
+        return canAccessPage(null, pageEntity);
+    }
+
+    @Override
+    public boolean canAccessPageFromConsole(PageEntity pageEntity) {
+        return canAccessPageFromConsole(null, pageEntity);
+    }
+
+
+    @Override
+    public boolean canAccessPageFromConsole(ApiEntity apiEntity, PageEntity pageEntity) {
+        if (canAccessPage(apiEntity, pageEntity)) {
+            return true;
+        } else {
+            return canEditDocumentation(apiEntity);
+        }
+    }
+
+    private boolean canEditDocumentation(ApiEntity api) {
+        boolean canEditDocumentation = false;
+        if (isAuthenticated()) {
+            MemberEntity member = membershipService.getUserMember(MembershipReferenceType.API, api.getId(), getAuthenticatedUsername());
```

Comment:
this method seems only related to membershipReferenceType.API, does it work well for platform level ?